File: advent_2021_d10.py
# -*- coding: utf-8 -*-
"""advent-2021-d10.ipynb
"""

f = open("input.txt", "r")
inputs = f.read().splitlines()
# print(inputs)

scores = {")": 3, "]": 57, "}": 1197, ">": 25137}
openers = {"(": ")", "[": "]", "{": "}", "<": ">"}
closers = { ")": "(", "]": "[", "}": "{", ">": "<"}

opened = list()
errors = list()
for input in inputs:
  for i in range(len(input)):
    currentChar = input[i]
    if currentChar in openers:
      opened.append(currentChar)
      continue
    elif currentChar in closers:
      if opened[len(opened)-1] == closers[currentChar]:
        opened = opened[:len(opened)-1]
        continue
      else:
        print(currentChar)
        # errors.append(currentChar)
        break

score = 0
for i in range(len(errors)):
  score += scores[errors[i]]
print("Total Syntax Scoring:", score)

File: advent-2021-d10.go
package main

import "fmt"

func main() {

	opened := []string{}
	errors := map[string]int{}

	for _, input := range inputs {
		for i := 0; i < len(input); i++ {
			currentChar := string(input[i])
			if _, isOpener := openers[currentChar]; isOpener {
				opened = append(opened, currentChar)
				continue
			} else if _, isCloser := closers[currentChar]; !isCloser {
				continue
			}
			if opened[len(opened)-1] == closers[currentChar] {
				opened = opened[:len(opened)-1]
				continue
			} else {
				errors[currentChar]++
				break
			}
		}
	}
	var total int = 0
	for err, times := range errors {
		fmt.Printf("\n %s: %d", err, times)
		total += scores[err] * times
	}
	fmt.Printf("\ntotal: %d", total)
}

var closers = map[string]string{
	")": "(",
	"]": "[",
	"}": "{",
	">": "<",
}

var openers = map[string]string{
	"(": ")",
	"[": "]",
	"{": "}",
	"<": ">",
}

var scores = map[string]int{
	")": 3,
	"]": 57,
	"}": 1197,
	">": 25137,
}

var inputs = []string{
	"[({(<(())[]>[[{[]{<()<>>",
	"[(()[<>])]({[<{<<[]>>(",
	"{([(<{}[<>[]}>{[]{[(<()>",
	"(((({<>}<{<{<>}{[]{[]{}",
	"[[<[([]))<([[{}[[()]]]",
	"[{[{({}]{}}([{[{{{}}([]",
	"{<[[]]>}<{[{[{[]{()[[[]",
	"[<(<(<(<{}))><([]([]()",
	"<{([([[(<>()){}]>(<<{{",
	"<{([{{}}[<[[[<>{}]]]>[]]",
}
